### PR TITLE
Add mixture of diag normals

### PR DESCRIPTION
**Pull request by martinjankowiak**
follows upon #1189 and #1255. last of the pathwise mixture distributions i intend to add at this point.

## Tested

- [x] tested extensively on internal experiments
- [x] add to existing `test_mean_gradient` test
- [x] add a simple hand-computed log prob test
- [x] add a test using model-side enumeration (following [test_elbo_zip](https://github.com/uber/pyro/blob/a156ca14795769d605ba3d836b2c931e84da2ba8/tests/infer/test_enum.py#L2754))
- ~~add a test of the continuity equation (following [test_dirichlet_tangent_field](https://github.com/pytorch/pytorch/blob/ef15bb8/test/test_distributions.py#L2529))~~

### CODE REVIEW

File: pyro/distributions/diag_normal_mixture.py [hunk 1]
```diff
@@ -0,0 +1,181 @@
...
+        mu_cd /= mu_cd_norm.unsqueeze(-1)  # b c d i
+        diagonals = torch.LongTensor(range(K))
+        mu_cd[..., diagonals, diagonals, :] = 0.0
+
+        mu_ll_cd = (locs.unsqueeze(-2) * mu_cd).sum(-1)  # b c d
+        z_ll_cd = (z.unsqueeze(-2).unsqueeze(-2) * mu_cd).sum(-1)  # l b c d
+        z_perp_cd = z.unsqueeze(-2).unsqueeze(-2) - z_ll_cd.unsqueeze(-1) * mu_cd  # l b c d i
+        z_perp_cd_sqr = torch.pow(z_perp_cd, 2.0).sum(-1)  # l b c d
+
+        shift_indices = torch.LongTensor(range(dim)) - 1
```
Comment (jpchen) [inline pyro/distributions/diag_normal_mixture.py]: this wont work for cuda. maybe use `z.new_tensor().long()` ?

File: pyro/distributions/diag_normal_mixture.py [hunk 2]
```diff
@@ -0,0 +1,181 @@
...
+        mu_cd_g = (g.unsqueeze(-2) * mu_cd).sum(-1)  # l b c d
+        v_cd *= -mu_cd_g * pis.unsqueeze(-2) * 0.5 * np.power(2.0 * math.pi, -0.5 * (dim - 1))  # l b c d
+        v_cd = pis * sum_leftmost(v_cd.sum(-1) / q_tot, -1 - batch_dims)
+        logits_grad += v_cd
+
+        prefactor = pis.unsqueeze(-1) * q_j * g / q_tot.unsqueeze(-1)
+        locs_grad = sum_leftmost(prefactor, -2 - batch_dims)
+        scales_grad = sum_leftmost(prefactor * z_tilde, -2 - batch_dims)
+
+        return locs_grad, scales_grad, logits_grad, None, None, None
```
Comment (jpchen) [inline pyro/distributions/diag_normal_mixture.py]: what are the last 3 fields in the tuple supposed to be?

File: pyro/distributions/diag_normal_mixture.py (hunk 1 above)
Comment (martinjankowiak) [reply pyro/distributions/diag_normal_mixture.py]: thanks i'll look into it...

File: pyro/distributions/diag_normal_mixture.py (hunk 2 above)
Comment (martinjankowiak) [reply pyro/distributions/diag_normal_mixture.py]: this is how you tell pytorch that particular inputs to a function don't have gradients

File: pyro/distributions/diag_normal_mixture.py (hunk 1 above)
Comment (fritzo) [reply pyro/distributions/diag_normal_mixture.py]: I think you'll want
```py
shift_indices = z.new_empty((dim,), dtype=torch.long)
torch.arange(dim, out=shift_indices)
```

File: pyro/distributions/diag_normal_mixture.py [hunk 3]
```diff
@@ -0,0 +1,181 @@
...
+
+        locs_tilde = locs / scales  # b j i
+        sigma_0 = torch.min(scales, -2, keepdim=True)[0]  # b 1 i
+        z_shift = (z.unsqueeze(-2) - locs) / sigma_0  # l b j i
+        z_tilde = z.unsqueeze(-2) / scales - locs_tilde  # l b j i
+
+        mu_cd = locs.unsqueeze(-2) - locs.unsqueeze(-3)  # b c d i
+        mu_cd_norm = torch.pow(mu_cd, 2.0).sum(-1).sqrt()  # b c d
+        mu_cd /= mu_cd_norm.unsqueeze(-1)  # b c d i
+        diagonals = torch.LongTensor(range(K))
```
Comment (fritzo) [inline pyro/distributions/diag_normal_mixture.py]: See below for cuda-compatible `arange` usage.

File: pyro/distributions/diag_normal_mixture.py [hunk 4]
```diff
@@ -0,0 +1,181 @@
...
+        z_shift_cumsum = torch.pow(z_shift, 2.0)
+        z_shift_cumsum = z_shift_cumsum.sum(-1, keepdim=True) - torch.cumsum(z_shift_cumsum, dim=-1)  # l b j i
+        z_tilde_cumsum = torch.cumsum(torch.pow(z_tilde, 2.0), dim=-1)  # l b j i
+        z_tilde_cumsum = torch.index_select(z_tilde_cumsum, -1, shift_indices)
+        z_tilde_cumsum[..., 0] = 0.0
+        r_sqr_ji = z_shift_cumsum + z_tilde_cumsum  # l b j i
+
+        log_scales = torch.log(scales)  # b j i
+        epsilons_sqr = torch.pow(z_tilde, 2.0)  # l b j i
+        log_qs = -0.5 * epsilons_sqr - 0.5 * np.log(2.0 * np.pi) - log_scales  # l b j i
```
Comment (fritzo) [inline pyro/distributions/diag_normal_mixture.py]: Use `math.log(2.0 * math.pi)` to avoid unnecessary numpy depencency (and finicky numpy float datatypes).

File: pyro/distributions/diag_normal_mixture.py [hunk 5]
```diff
@@ -0,0 +1,181 @@
...
+        q_j = torch.exp(log_q_j)  # l b j 1
+        q_tot = (pis * q_j.squeeze(-1)).sum(-1)  # l b
+        q_tot = q_tot.unsqueeze(-1)  # l b 1
+
+        root_two = math.sqrt(2.0)
+        shift_log_scales = log_scales[..., shift_indices]
+        shift_log_scales[..., 0] = 0.0
+        sigma_products = torch.cumsum(shift_log_scales, dim=-1).exp()  # b j i
+
+        reverse_indices = torch.LongTensor(range(dim - 1, -1, -1))
```
Comment (fritzo) [inline pyro/distributions/diag_normal_mixture.py]: See above for cuda-compatible `torch.arange` invocation.

Comment (fritzo) [reply pyro/distributions/diag_normal_mixture.py]: If `arange` didn't work, maybe `sigma_0.new_tensor(range(dim - 1, -1, -1), dtype=torch.long)` would work?
